### DISCUSSION TITLE
How do I configure a script with project, supporting files, etc so it can be run from the PATH?

### DISCUSSION THREAD

**Discussion by TuggyNE** [babashka/babashka#1525] [Q&A] [answered]
I have several scripts I'm writing that are intended to be used like normal commands. The shebang is easy enough: `#!/usr/bin/env bb`. But they rely on `cprop/cprop`, which introduces data file dependencies (bb.edn and config.edn) that should be in the same directory as the script. bb.edn looks like this:
```clojure
{:paths ["."]
 :deps {org.babashka/spec.alpha {:git/url "https://github.com/babashka/spec.alpha"
                                 :git/sha "1a841c4cc1d4f6dab7505a98ed2d532dd9d56b78"}
        org.clj-commons/clj-http-lite {:mvn/version "1.0.13"}
        com.cognitect/anomalies {:mvn/version "0.1.12"}
        cprop/cprop {:mvn/version "0.1.19"}
        failjure/failjure {:mvn/version "2.2.0"}
        zprint/zprint {:mvn/version "1.2.5"}}}
```
This works fine if I'm running it from the same directory, e.g. `./script.clj`. But it fails (`Message:  Could not find namespace: cprop.core.`) if I run from PATH, e.g. (in another directory) `script.clj`. So clearly bb.edn is not even being read. How should I fix this?

Unlike in https://clojureverse.org/t/help-utilizing-babashka-tasks-globally/8026/2, there are multiple bb.edn files (for good reason) in different directories, all of which are on the PATH.

**Comment by borkdude** [comment]
@TuggyNE This is where bbin comes in:

https://github.com/babashka/bbin

**Reply by TuggyNE** [reply]
Thanks! I'll have to try this, although it will take a bit before I can report back.

**Reply by TuggyNE** [reply]
I can't figure out what incantation is necessary to actually get `bbin` to recognize and make use of an appropriate deps.edn.

* `bbin install path/to/containing/dir/` (with script and bb.edn there) requires bbin-specific customization to bb.edn to get back default behavior, which is irritating
* When that's done, `bbin install path/to/containing/dir/` doesn't work, because it didn't bother installing dependencies from bb.edn, even though it clearly _read_ the file
* `bbin install --local/root path/to/containing/dir/` just gives usage directions
* `bbin install --local/root path/to/containing/dir/ path/to/containing/dir/` gives a complicated error that suggests it's completely wrong
* `bbin install --local/root path/to/containing/dir/ path/to/containing/dir/script.clj` doesn't bother to check for deps.edn, so fails in the same way as just adding to the PATH
* `bbin install https://ssl-enabled-localhost/path/to/containing/dir` reports `Invalid script coordinates.
If you're trying to install from the filesystem, make sure the path actually exists.` 
* `bbin install https://ssl-enabled-localhost/path/to/containing/dir/script.clj`, yet again, installs just fine but fails to download any dependencies
* `--git/url` is unworkable for this case (since it requires a public repo-per-script, and neither of those criteria holds or will ever hold)

**Reply by borkdude** [reply]
@TuggyNE I think you might have to add `{:bbin/bin {your-tool {:main-opts ["-f" "script.clj"]}}}` to your `bb.edn` and then you can just run `bbin install .` in the project directory. Change the name of `your-tool` accordingly.

cc @rads This may be good feedback on the usability of bbin.

**Reply by borkdude** [reply]
@TuggyNE Another thing that's missing: you may have to move any dependencies to `deps.edn` and then refer to your local deps edn from the bb.edn:

In bb.edn:
```
{:deps {my/project {:local/root "."}}}
```

In `deps.edn`:

```
{:deps ...}
```

**Reply by borkdude** [reply]
@TuggyNE If you're not able to solve it, could maybe make a repro project, that I can clone locally? This way I can see better what should be done.

**Reply by TuggyNE** [reply]
@borkdude Looks like adding `:deps {script/script {:local/root "."}} :bbin/bin {script {:main-opts ["-f" "script.clj"]}}` does the job (and all of those are required, including the redundantly-qualified symbol in `:deps` and the unqualified symbol in `:bbin/bin`), but rewriting all the EDN files and also adding support for running bbin on all of them is probably more work than just setting up a simple wrapper script.

**Reply by borkdude** [reply]
OK, I'm evaluating the option that babashka should always consider the bb.edn relative to a script, in precedence over the bb.edn in the current working directory. That is the same behavior of Node.js and might make sense for scripting tools like bb. This would also simplify your problem.

**Reply by TuggyNE** [reply]
Yeah, I can't speak to any breaking changes (in that I am struggling to imagine any vaguely legitimate situation where two bb.edn files would need to be evaluated with the one in $PWD before the one in $PATH), but it would absolutely make more sense to me to have this changed.

**Comment by borkdude** [comment, marked as answer]
A solution without bbin is the following:

Write a bash wrapper script, e.g. 

```
#!/usr/bin/env bash

bb --config ~/my_project/bb.edn ~/my_project/script.clj
```

Make that bash wrapper script executable and put it on your path.

**Reply by borkdude** [reply]
@TuggyNE So this solved your problem, right?
